### PR TITLE
Support flash/flex/xformers/sage attention

### PR DESCRIPTION
**Pull request by a-r-r-o-w**
This PR adds support for various attention providers. Every `F.scaled_dot_product_attention` call is hijacked with a custom attention dispatcher.

For training:
- Model uses attention masks:
  - `flash_varlen`, `flex` (must have attention head dim as a power of 2), `native`.
  - Additionally, for debugging purposes, `_native_cudnn` and `_native_efficient`
- Model does not use attention masks:
  - All of the above + `flash`, `xformers`
  - Additionally, for debugging purposes, `_native_flash`

For inference:
- Model uses attention masks: Same as training with attention masks + `sage_varlen`
- Model does not use attention masks: all of the above

Currently, setting the attention backend can only be done by modifying environment flags:  `FINETRAINERS_ATTN_PROVIDER="flash_varlen"`. More precise control over applying different attention for different parts of the model is planned in the future. This is required because of some unfortunate model implementations that are not compatible with existing attention providers: for example, some VAEs with attention layers have weird head dim and sequence length configurations/masks that will error out with, say, flash-attn.

Each attention implementation has some validation checks that can be enabled by setting `FINETRAINERS_ATTN_CHECKS="1"` (or a similar true value). These basic checks are helpful for debugging purposes.

### Linux

This PR has only been tested on Ubuntu 22.04. Two different setups were tested:
- 4x A100 with pre-existing CUDA 12.2 (Driver version: 535.129.03)
- 2x 4090 with pre-existing CUDA 12.6 (Driver version: 560.35.05)

1. To use `flash` and `flash_varlen`, install [flash-attn](https://github.com/Dao-AILab/flash-attention):
  - Verify that you have atleast CUDA 12.0 or above. The easiest way to check this is by looking at the output of `nvidia-smi` or `nvcc --version`
  - `pip install packaging ninja`
  - `pip install flash-attn --no-build-isolation`
  - Make sure you have atleast `flash-attn>=2.6.3` (the requirements will be lowered eventually as more tests are done)

2. To use `sage` and `sage_varlen`, install [SageAttention](https://github.com/thu-ml/SageAttention):
  - Verify you match the CUDA version requirements as mentioned on the SageAttention repo
  - `pip install git+https://github.com/thu-ml/SageAttention`
  - Make sure you have atleast `sageattention>=2.1.1`. Currently, a source install is a must but the requirements will be lowered eventually with more testing

3. To use `xformers`, install [xformers](https://github.com/facebookresearch/xformers):
  - On my device with 4090, the following worked (tested with Pytorch 2.6 only): `pip3 install -U xformers --index-url https://download.pytorch.org/whl/cu126`

### Windows

This PR has not been tested on Windows. It is expected to work with the same setup instructions as Linux if using WSL. (TODO: point to available online guides on how to setup)

Any help for improving Windows support would be very highly appreciated!

Other changes:
- Move ConfigMixin to utils since it will be commonly used outside of specifying additional trainer args. This also gets rid of a few possibilities of circular imports, due to which some imports were done only when `TYPE_CHECKING`
- Cleanup import code related to package availability and versions

### CODE REVIEW

**Comment by a-r-r-o-w** [comment]
To run different attention providers for sanity check:

<details>
<summary> train.sh </summary>

```bash
#!/bin/bash

set -e -x

# export TORCH_LOGS="+dynamo,recompiles,graph_breaks"
# export TORCHDYNAMO_VERBOSE=1
export WANDB_MODE="offline"
export NCCL_P2P_DISABLE=1
export TORCH_NCCL_ENABLE_MONITORING=0
export FINETRAINERS_LOG_LEVEL="DEBUG"

# Finetrainers supports multiple backends for distributed training. Select your favourite and benchmark the differences!
# BACKEND="accelerate"
BACKEND="ptd"

# In this setting, I'm using 2 GPUs on a 4-GPU node for training
NUM_GPUS=2
CUDA_VISIBLE_DEVICES="2,3"

# Check the JSON files for the expected JSON format
TRAINING_DATASET_CONFIG="examples/training/sft/wan/3dgs_dissolve/training.json"
VALIDATION_DATASET_FILE="examples/training/sft/wan/3dgs_dissolve/validation.json"

# Depending on how many GPUs you have available, choose your degree of parallelism and technique!
DDP_1="--parallel_backend $BACKEND --pp_degree 1 --dp_degree 1 --dp_shards 1 --cp_degree 1 --tp_degree 1"
DDP_2="--parallel_backend $BACKEND --pp_degree 1 --dp_degree 2 --dp_shards 1 --cp_degree 1 --tp_degree 1"
DDP_4="--parallel_backend $BACKEND --pp_degree 1 --dp_degree 4 --dp_shards 1 --cp_degree 1 --tp_degree 1"
FSDP_2="--parallel_backend $BACKEND --pp_degree 1 --dp_degree 1 --dp_shards 2 --cp_degree 1 --tp_degree 1"
FSDP_4="--parallel_backend $BACKEND --pp_degree 1 --dp_degree 1 --dp_shards 4 --cp_degree 1 --tp_degree 1"
HSDP_2_2="--parallel_backend $BACKEND --pp_degree 1 --dp_degree 2 --dp_shards 2 --cp_degree 1 --tp_degree 1"

# Parallel arguments
parallel_cmd=(
  $DDP_2
)

# Model arguments
model_cmd=(
  --model_name "wan"
  --pretrained_model_name_or_path "Wan-AI/Wan2.1-T2V-1.3B-Diffusers"
)

# Dataset arguments
# Here, we know that the dataset size if about ~100 videos. Since we're using 2 GPUs, we precompute
# embeddings of 50 dataset items per GPU. Also, we're using a very small dataset for finetuning, so
# we are okay with precomputing embeddings once and re-using them without having to worry about disk
# space. Currently, however, every new training run performs precomputation even if it's not required
# (which is something we've to improve [TODO(aryan)])
dataset_cmd=(
  --dataset_config $TRAINING_DATASET_CONFIG
  --dataset_shuffle_buffer_size 1
)

# Dataloader arguments
dataloader_cmd=(
  --dataloader_num_workers 0
)

# Diffusion arguments
diffusion_cmd=(
  --flow_weighting_scheme "logit_normal"
)

# Training arguments
# We target just the attention projections layers for LoRA training here.
# You can modify as you please and target any layer (regex is supported)
training_cmd=(
  --training_type "lora"
  --seed 42
  --batch_size 1
  --train_steps 10
  --rank 32
  --lora_alpha 32
  --target_modules "blocks.*(to_q|to_k|to_v|to_out.0)"
  --gradient_accumulation_steps 1
  --gradient_checkpointing
  --checkpointing_steps 500
  --checkpointing_limit 2
  # --resume_from_checkpoint 3000
  --enable_slicing
  --enable_tiling
)

# Optimizer arguments
optimizer_cmd=(
  --optimizer "adamw"
  --lr 5e-5
  --lr_scheduler "constant"
  --beta1 0.9
  --beta2 0.99
  --weight_decay 1e-4
  --epsilon 1e-8
  --max_grad_norm 1.0
)

# Validation arguments
validation_cmd=(
  --validation_dataset_file "$VALIDATION_DATASET_FILE"
  --validation_steps 6
)

# Miscellaneous arguments
miscellaneous_cmd=(
  --tracker_name "finetrainers-debug"
  --output_dir "/raid/aryan/wan"
  --init_timeout 600
  --nccl_timeout 600
  --report_to "wandb"
)

# Attention provider arguments
# attention_provider_cmd=(
#   --attn_provider_training native
#   --attn_provider_validation native
# )
# attention_provider_cmd=(
#   --attn_provider_training flash_varlen
#   --attn_provider_validation native
# )
attention_provider_cmd=(
  --attn_provider_training xformers
  --attn_provider_validation xformers
)

# Execute the training script
if [ "$BACKEND" == "accelerate" ]; then

  ACCELERATE_CONFIG_FILE=""
  if [ "$NUM_GPUS" == 1 ]; then
    ACCELERATE_CONFIG_FILE="accelerate_configs/uncompiled_1.yaml"
  elif [ "$NUM_GPUS" == 2 ]; then
    ACCELERATE_CONFIG_FILE="accelerate_configs/uncompiled_2.yaml"
  elif [ "$NUM_GPUS" == 4 ]; then
    ACCELERATE_CONFIG_FILE="accelerate_configs/uncompiled_4.yaml"
  elif [ "$NUM_GPUS" == 8 ]; then
    ACCELERATE_CONFIG_FILE="accelerate_configs/uncompiled_8.yaml"
  fi
  
  accelerate launch --config_file "$ACCELERATE_CONFIG_FILE" --gpu_ids $CUDA_VISIBLE_DEVICES train.py \
    "${parallel_cmd[@]}" \
    "${model_cmd[@]}" \
    "${dataset_cmd[@]}" \
    "${dataloader_cmd[@]}" \
    "${diffusion_cmd[@]}" \
    "${training_cmd[@]}" \
    "${optimizer_cmd[@]}" \
    "${validation_cmd[@]}" \
    "${miscellaneous_cmd[@]}" \
    "${attention_provider_cmd[@]}"

elif [ "$BACKEND" == "ptd" ]; then

  export CUDA_VISIBLE_DEVICES=$CUDA_VISIBLE_DEVICES
  
  torchrun \
    --standalone \
    --nnodes=1 \
    --nproc_per_node=$NUM_GPUS \
    --rdzv_backend c10d \
    --rdzv_endpoint="localhost:0" \
    train.py \
      "${parallel_cmd[@]}" \
      "${model_cmd[@]}" \
      "${dataset_cmd[@]}" \
      "${dataloader_cmd[@]}" \
      "${diffusion_cmd[@]}" \
      "${training_cmd[@]}" \
      "${optimizer_cmd[@]}" \
      "${validation_cmd[@]}" \
      "${miscellaneous_cmd[@]}" \
      "${attention_provider_cmd[@]}"
fi

echo -ne "-------------------- Finished executing script --------------------\n\n"
```

</details>

<details>
<summary> training.json </summary>

```json
{
  "datasets": [
    {
      "data_root": "finetrainers/3dgs-dissolve",
      "dataset_type": "video",
      "id_token": "3DGS_DISSOLVE",
      "video_resolution_buckets": [
        [49, 480, 832]
      ],
      "reshape_mode": "bicubic",
      "remove_common_llm_caption_prefixes": true
    }
  ]
}
```

</details>

<details>
<summary> validation.json </summary>

```json
{
  "data": [
    {
      "caption": "A spacecraft, rendered in a 3D appearance, ascends into the night sky, leaving behind a trail of fiery exhaust. As it climbs higher, the exhaust gradually transforms into a burst of red sparks, creating a dramatic and dynamic visual effect against the dark backdrop.",
      "image_path": null,
      "video_path": null,
      "num_inference_steps": 2,
      "height": 480,
      "width": 832,
      "num_frames": 49
    }
  ]
}
```

</details>

Relevant params:
- `attn_provider_training`: only affects transformer forward/backward for now
- `attn_provider_inference`: affects all modeling components. Make sure to verify a dummy run first before starting anything big (for example, you might error out in VAE encode/decode if it contains attention layers that have >256 head dims and you're using flash_varlen or sage_varlen)

File: finetrainers/models/attention_dispatch.py [hunk 1]
```diff
@@ -0,0 +1,905 @@
...
+    seqlens_q = torch.full((batch_size,), seq_len_q, dtype=torch.int32, device=device)
+    if attn_mask is None:
+        seqlens_k = torch.full((batch_size,), seq_len_kv, dtype=torch.int32, device=device)
+    else:
+        seqlens_k = attn_mask.sum(dim=1, dtype=torch.int32)
+    cu_seqlens_q = torch.zeros(batch_size + 1, dtype=torch.int32, device=device)
+    cu_seqlens_k = torch.zeros(batch_size + 1, dtype=torch.int32, device=device)
+    cu_seqlens_q[1:] = torch.cumsum(seqlens_q, dim=0)
+    cu_seqlens_k[1:] = torch.cumsum(seqlens_k, dim=0)
+    max_seqlen_q = seqlens_q.max().item()
```
Comment (a-r-r-o-w) [inline finetrainers/models/attention_dispatch.py:276]: This is a graph break at every layer that calls `F.scaled_dot_product_attention`. Ideally, we should need to do the preparation only once (at the first layer or before it) and pass these values, as they don't change between transformer blocks. This will cause unnecessary slowdown in both compiled and non-compiled case.

File: finetrainers/models/attention_dispatch.py [hunk 2]
```diff
@@ -0,0 +1,905 @@
...
+    out = out.unflatten(0, (batch_size, -1)).permute(0, 2, 1, 3)  # .contiguous()
+
+    return out
+
+
+@_AttentionProviderRegistry.register(
+    AttentionProvider.FLEX,
+    constraints=[_check_attn_mask_or_causal, _check_device, _check_shape],
+)
+def _native_flex_attention(
```
Comment (a-r-r-o-w) [inline finetrainers/models/attention_dispatch.py:460]: Flex attention is not useful without compilation. Will be much much slower
